index.rst

```diff
@@ -1,61 +1,499 @@
+This is related to the long-running middleware goal of better integrating ``pipetask`` and BPS.
+It will probably also involve carving out a third aspect of butler (a new sibling to ``Registry`` and ``Datastore``) for ``QuantumGraph`` and provenance files, because
+
+- like ``Datastore``, this aspect would be backed by files in a shared filesystem or object store (but not necessarily the same filesystem or bucket as an associated ``Datastore``);
+- like ``Registry``, the new aspect would provide descriptive and organizational metadata for datasets, rather than hold datasets themselves, and after execution its content would be completely loaded into the ``Registry``.
+
+The full high-level design will be the subject of a future technote, but an early sketch can be found `in Confluence <https://confluence.lsstcorp.org/display/DM/Saving+per-Quantum+provenance+and+propagating+nothing-to-do+cases%2C+and+The+Future>`_.
+A key point is that ultimately we want inserting datasets back into a permanent registry to be a deferred, final step in all modes of execution, not just BPS, allowing that step to transform file-based data in various ways before it is loaded into a database.
+
+Recording provenance only when using BPS (and relying on it to manage the ``QuantumGraph`` and provenance files) in the interim seems like a good first step.
+Extending the design to include non-BPS processing may take time, but we do not anticipate it changing what happens at a low level, or the appearance of persisted provenance information.
+
+.. _failures_and_retries:
+
+Failures and retries
+--------------------
+
+When a quantum fails during execution, we will often want to try running it again.
+This may happen immediately via workflow system functionality or manually via a resubmission.
+The outputs of any such failure are considered unreliable, and we don't want to proceed with downstream processing even if some outputs were written before the failure (not producing all - or even any - outputs is not necessarily considered a failure).
+We want actual failures to be included in provenance, and their datasets included in the main data repository, until/unless they are explicitly deleted after any analysis of the failure is complete.
+This leads to a conflict: running a failed quantum again would lead to two ``quantum`` rows with the same UUID, data ID, and ``RUN``, and analoguous conflicts for any output datasets that were written before the failure occurred or in spite of it (e.g. logs written by the execution system after it catches an exception).
+
+The possibility of immediate retries performed by the workflow system means we cannot assign the retry quantum and datasets new UUIDs, because this would prevent downstream quanta from being able to find their inputs.
+Instead, we have to assign new UUIDs (and a new RUN collection) to the failures, and move all datasets that were already written to permanent storage out of the way.
```

Comment:
"move all datasets that were already written to permanent storage out of the way"
This seems time consuming as well as possibly problematic if we're talking about the retry starting after the merge has been done (e.g., when will Rucio be told about the files)